src/Application/Kernel.php

```diff
@@ -8,39 +8,38 @@
 use RecursiveIteratorIterator;
 use ReflectionClass;
 use Tempest\AppConfig;
-use Tempest\Application\Exceptions\KernelException;
 use Tempest\Container\Container;
 use Tempest\Container\GenericContainer;
 use Tempest\Database\PDOInitializer;
-use Tempest\Discovery\DiscoveryLocation;
 use Tempest\Http\RequestInitializer;
 use Tempest\Http\RouteBindingInitializer;
 use Tempest\Http\ServerInitializer;
-use function Tempest\path;
 use Throwable;
 
 final readonly class Kernel
 {
     public function __construct(
-        private string $root,
+        public string $root,
```

Comment:
Why are we changing the visibility here? If we really need global access, it should probably be in the config file.

Reply:
It's the only part of this process i couldn't find a clean way to get into the Bootstrap logic, where Discovery needs it. i'm fine moving it somewhere else. Which config file are you referring to?

Reply:
I think I would be inclined to move this to `AppConfig` over exposing it as a public property.

Reply:
i considered that, but i wasn't sure i like adding another parameter to its constructor. It already feels like anything we want to hold in configuration has to be added as a constructor param. i'm not against it, that's just where i fell when i was examining it.

Reply:
@brendt / @aidan-casey 

i agree with Aidan that `$root` shouldn't _need_ to be exposed publicly. But there's no easy way to get it from Kernel to Container or from Kernel to Bootstraps (or anywhere else). We can add it to AppConfig as a new constructor parameter, but i'm not loving that either. We could add a function to add arbitrary data to AppConfig, but... You get the idea.

Any suggestions?

Reply:
I think we need to determine _what_ we are bootstrapping here. I am of the mind that we are not bootstrapping the `container`, but the application. In that case, the bootstrapper should _only_ be receiving an instance of the application and not have full control over the container. This will also guard us against circular dependencies and an inappropriate use of bootstrappers.

I recommend looking through how Symfony and Laravel use these for some inspiration.

Reply:
i'll take some time to review those and a couple others i've worked with recently. i'll try to come up with options soon. If anyone else has suggestions in the meantime, please post.

Reply:
(Originally posted in the wrong thread....)

Here's what i see happening in Spiral:

The Kernel is constructed by passing an array of named directories:

[Tutorial app example](https://github.com/spiral/app/blob/master/app.php#L20)
[Method declaration](https://github.com/spiral/framework/blob/master/src/Boot/src/AbstractKernel.php#L109)

The array becomes a Directories object in the Container:
[Kernel constructor](https://github.com/spiral/framework/blob/master/src/Boot/src/AbstractKernel.php#L85)

It may be worth noting the Kernel builds the Container in Kernel::create() and passes it to the constuctor... i guess that permits other approaches to construct the Kernel with a different Container implementation, where Tempest's Kernel wants it done with the Tempest Container.

For future consideration, we may want to refactor to permit other Containers, but not today.

If there's no opposition, i'm going to add a similar Directories class that the Container will be able to resolve, and take care of this exposure issue. i was already headed somewhat in that direction earlier, but it felt like overkill. But after more thought it makes sense - other application types may need to keep track of various file system resources.

Reply:
@aidan-casey's response:

---

So in reviewing, I actually think Spiral is a pretty poor example of a bootstrapper implementation.

Laravel, I think, gives a better perspective into what a bootstrapper should be doing. See [here](https://github.com/laravel/framework/blob/698c74b8465daa518cf3954eea133e4e6e3f47d7/src/Illuminate/Foundation/Http/Kernel.php#L42) for example.

A bootstrapper isn't really getting hands on with the application. It is getting and providing the data necessary to help setup the kernel so that the actual application can load. Much like a computers boot process.

To me, there are a few things that are concerning to me about this PR:

1. It seems to have taken a turn away from kernel bootstrapping and toward setting up dependencies in the container. We basically already have this in initializers, so I'm not entirely sure what we are gaining.
2. The scope of the PR keeps changing/getting added to.

I don't say that to discourage you from going down this path, but only to encourage you to target one specific problem to tackle here so that we can merge this specific PR quickly and look to tackle other issues in future PRs. It would be helpful to have a clear argument for what the issue is that we are trying to solve, and how bootloaders/bootstrappers are going to do that for us.

Reply:
Put in its simplest form, this PR intends to move Discovery and Config out of the Kernel itself as a precursor to implementing discovery cache.


The changes in this PR are targeting just that one objective. There were some unrelated changes that you rightfully pointed out should be put back to limit scope, so those reversions are upcoming. But the basic change set remains:

1, Move Discovery-related methods out of the Kernel
2. Move `initConfig()` out of the Kernel
3. Provide the new "Bootstrap" classes made from 1 & 2 to the Kernel so it can construct them and execute the original logic

---

The outline of changes i offered seemed reasonable for doing that. The extractions don't really qualify as "Initializers" as they're not establishing application resources - they're discovering dependencies that will establish those resources...

i looked over the link to Illuminate's Http Kernel, but i honestly don't see it behaving any differently from Spiral. And to be clear, i'm not a Spiral evangelist by any means. i've just been working with it lately.

As to the concerns about the PR:

> It seems to have taken a turn away from kernel bootstrapping and toward setting up dependencies in the container. We basically already have this in initializers, so I'm not entirely sure what we are gaining.

i don't see it this way. The fundamental changes here just pull the Discovery/Config functions out to their own "tasks" - Bootstraps. They still only do what they were doing before, and no new functionality was added to the Kernel itself.

>The scope of the PR keeps changing/getting added to.

The scope changes have been attempting to address concerns raised. If we want to limit that scope creep, we can always set up issues to attack them later. My offer to add the Directories class was an attempt to do this somewhat elegantly, but it doesn't "fix" anything other than moving a public property into an object in the Container.

The `$root` visibility can be solved in a few different ways (maybe more):
1. `public` - minimal change to satisfy what Discovery/Config need
    * Leaves a bit of code smell, although it _is_ still within an object in the Container
2. Add to `AppConfig`
    * Expands PR scope via one of the following:
        a. Requires expanding AppConfig::__construct() signature for another parameter,
        b. OR adding logic to maintain arbitrary data in the AppConfig object
5. Add to Container
    * Requires making it resolvable (Directories class or similar that gets added to Container)
    * Expands PR scope

---

In my opinion, the visibility change is the minimal change that achieves the goal without expanding scope more than necessary. The extraction is functional, and we should be able to move forward with the cache implementation.

We can easily revisit what to do with it in a future adjustment.

Or, maybe what we're calling Bootstrapping for purposes of this change should just remain internal in the Kernel, and as you say, let "Initializers" be the mechanism that does anything else.

@brendt and @aidan-casey
i'll catch up this branch, and let you decide what happens to the PR. If there are specific ways you want it done differently, please let me know.